api/v1alpha1/serviceintentions_types.go

```diff
@@ -296,12 +333,34 @@ func (in IntentionPermissions) validate(path *field.Path) field.ErrorList {
 
 func (in *IntentionHTTPPermission) validate(path *field.Path) field.ErrorList {
 	var errs field.ErrorList
+	if (in.PathExact != "" && in.PathPrefix != "") || (in.PathExact != "" && in.PathRegex != "") || (in.PathPrefix != "" && in.PathRegex != "") {
+		errs = append(errs, field.Invalid(path, in, `At most only one of pathExact, pathPrefix, or pathRegex may be configured.`))
+	}
 	if invalidPathPrefix(in.PathPrefix) {
 		errs = append(errs, field.Invalid(path.Child("pathPrefix"), in.PathPrefix, `must begin with a '/'`))
 	}
 	if invalidPathPrefix(in.PathExact) {
 		errs = append(errs, field.Invalid(path.Child("pathExact"), in.PathExact, `must begin with a '/'`))
 	}
+	for i, method := range in.Methods {
+		methods := []string{"GET", "HEAD", "POST", "PUT", "DELETE", "CONNECT", "OPTIONS", "TRACE", "PATCH"}
+		if !sliceContains(methods, strings.ToUpper(method)) {
+			errs = append(errs, field.Invalid(path.Child("methods").Index(i), method, notInSliceMessage(methods)))
+		}
+	}
+	errs = append(errs, in.Header.validate(path.Child("header"))...)
+	return errs
+}
+func (in IntentionHTTPHeaderPermissions) validate(path *field.Path) field.ErrorList {
+	var errs field.ErrorList
+	for i, permission := range in {
+		if (permission.Regex != "" && permission.Suffix != "") || (permission.Regex != "" && permission.Prefix != "") || (permission.Regex != "" && permission.Exact != "") ||
+			(permission.Regex != "" && permission.Present) || (permission.Suffix != "" && permission.Prefix != "") || (permission.Suffix != "" && permission.Exact != "") ||
+			(permission.Suffix != "" && permission.Present) || (permission.Prefix != "" && permission.Exact != "") ||
+			(permission.Prefix != "" && permission.Present) || (permission.Exact != "" && permission.Present) {
+			errs = append(errs, field.Invalid(path.Index(i), in[i], `At most only one of exact, prefix, suffix, regex, or present may be configured.`))
```

Comment:
Could we count how many strings are non-empty and error if count != 1?

Reply:
I like this idea, but I think we'd need to error if count > 1 since 0 is also a valid count. It seems that consul validates it in this way too.

Reply:
Good catch. Updating this check.